mlserver/batching/adaptive.py

```diff
@@ -0,0 +1,94 @@
+import time
+import asyncio
+
+from asyncio import Future, Queue, wait_for
+from typing import AsyncIterator, Awaitable, Dict, List
+
+from ..model import MLModel
+from ..types import (
+    InferenceRequest,
+    InferenceResponse,
+)
+
+from .requests import BatchedRequests
+
+
+class AdaptiveBatcher:
+    def __init__(self, model: MLModel):
+        self._model = model
+
+        self._max_batch_size = model.settings.max_batch_size
+        self._max_batch_time = model.settings.max_batch_time
+
+        # Save predict function before it gets decorated
+        self._predict_fn = model.predict
+        self._requests: Queue[InferenceRequest] = Queue(maxsize=self._max_batch_size)
+        self._async_responses: Dict[str, Future[InferenceResponse]] = {}
+        self._batching_task = None
+
+    async def predict(self, req: InferenceRequest) -> InferenceResponse:
+        await self._queue_request(req)
+        self._start_batcher_if_needed()
+        return await self._wait_response(req)
+
+    async def _queue_request(
+        self,
+        req: InferenceRequest,
+    ) -> Awaitable[InferenceResponse]:
+        await self._requests.put(req)
+
+        loop = asyncio.get_running_loop()
+        async_response = loop.create_future()
+
+        # TODO: What happens if ID is None?
+        self._async_responses[req.id] = async_response  # type: ignore
+
+        return async_response
+
+    async def _wait_response(self, req: InferenceRequest) -> InferenceResponse:
+        # TODO: What happens if ID is None?
+        pred_id = req.id
+        async_response = self._async_responses[pred_id]  # type: ignore
+
+        response = await async_response
+
+        del self._async_responses[pred_id]  # type: ignore
+        return response
+
+    def _start_batcher_if_needed(self):
+        if self._batching_task is not None:
+            if not self._batching_task.done():
+                # If task hasn't finished yet, let it keep running
+                return
+
+        # TODO: If _batching_task crashes, cancel all async responses
+        self._batching_task = asyncio.create_task(self._batcher())
+
+    async def _batcher(self):
+        async for batched in self._batch_requests():
```

Comment:
Good point. That's my impression, but I'm not 100% sure. We should write a quick test to ensure that.

Reply:
I had a look, and turns out that it doesn't run different iterations concurrently. Even if all the for-loop body is encapsuled in a function, it will still run each iteration sequentially. One potential fix would be to use `asyncio.gather`, however we won't necessarily have all batches at the same time.

Instead, the PR should now leverage `asyncio.create_task()` and callbacks to ensure that the batches get scheduled for prediction immediately and can run concurrently.